util/git/client.go

```diff
@@ -477,11 +478,26 @@ func (m *nativeGitClient) Checkout(revision string, submoduleEnabled bool) error
 }
 
 func (m *nativeGitClient) getRefs() ([]*plumbing.Reference, error) {
+	// Prevent an additional get call to cache if we know our state isn't stale
+	needsUnlock := true
 	if m.gitRefCache != nil && m.loadRefFromCache {
 		var res []*plumbing.Reference
-		if m.gitRefCache.GetGitReferences(m.repoURL, &res) == nil {
+		updateCache, lockId, err := m.gitRefCache.GetOrLockGitReferences(m.repoURL, &res)
+		if !updateCache && err == nil {
+			// Valid value already in cache
 			return res, nil
+		} else if !updateCache && err != nil {
+			// Error getting value from cache
+			log.Debugf("Error getting git references from cache: %v", err)
+			return nil, err
 		}
+		// Defer a soft reset of the cache lock, if the value is set this call will be ignored
+		defer func() {
+			if needsUnlock {
+				err := m.gitRefCache.UnlockGitReferences(m.repoURL, lockId)
+				log.Debugf("Error unlocking git references from cache: %v", err)
+			}
+		}()
```

Comment:
I am trying to understand the locking logic.
`updateCache` is the same as `lockId != ""`. If the cache is already locked by another client, this method should not return the lock ID of the other client

Should `updateCache` be removed? I think the name updateCache represent the actions to do instead of the result of the function.
Should `needsUnlock` be set to the value of `hasLock`? If the current code does not have the lock, it seems unnecessary to call the unlock method.

```suggestion
	        lockId, err := m.gitRefCache.GetOrLockGitReferences(m.repoURL, &res)
	        hasLock := lockId != ""
		if !hasLock && err == nil {
			// Valid value already in cache
			return res, nil
		} else if !hasLock && err != nil {
			// Error getting value from cache
			log.Debugf("Error getting git references from cache: %v", err)
			return nil, err
		}
		needsUnlock = hasLock
		// Defer a soft reset of the cache lock, if the value is set this call will be ignored
		defer func() {
			if needsUnlock && lockId != "" {
				err := m.gitRefCache.UnlockGitReferences(m.repoURL, lockId)
				log.Debugf("Error unlocking git references from cache: %v", err)
			}
		}()
```

Reply:
I see the confusion here with the naming `updateCache` is verification of ownership from outside the lock call, it isn't equal to lockId != "", it's equal to remote lock id == local lock id. `needsUnlock` is a useless variable here though. `lockId` is just returned so it can be referenced during the unlock. 

For the defer piece, that will only be called when the local is the owner e.g. `updateCache` is true or there is an error in which case it is still treated as the owner.

I'll find some better names for these variables

Reply:
Renamed the variables but otherwise did not make additional changes here

Reply:
I like `isLockOwner` 👍